src/core/lib/security/transport/secure_endpoint.cc

```diff
@@ -410,8 +411,26 @@ static void endpoint_write(grpc_endpoint* secure_ep, grpc_slice_buffer* slices,
 
     if (ep->zero_copy_protector != nullptr) {
       // Use zero-copy grpc protector to protect.
-      result = tsi_zero_copy_grpc_protector_protect(ep->zero_copy_protector,
-                                                    slices, &ep->output_buffer);
+      grpc_slice_buffer_reset_and_unref_internal(
+          &ep->tmp_frame_size_clip_buffer);
+      result = TSI_OK;
+      // Break the input slices into chunks of size = max_frame_size and call
+      // tsi_zero_copy_grpc_protector_protect on each chunk. This ensures that
+      // the protector cannot create frames larger than the specified
+      // max_frame_size.
+      while (slices->length > static_cast<size_t>(max_frame_size) &&
+             result == TSI_OK) {
+        grpc_slice_buffer_move_first(slices,
+                                     static_cast<size_t>(max_frame_size),
+                                     &ep->tmp_frame_size_clip_buffer);
+        result = tsi_zero_copy_grpc_protector_protect(
```

Comment:
The answer is probably yes, but just to clarify - the new protected chunk will be appended to output buffer?

Reply:
Yeah the implementations we have only append to the output buffer. They don't reset the output buffer before appending.